Solutions/Corvus.Identity.ManagedServiceIdentity.ClientAuthentication/Microsoft/Extensions/DependencyInjection/ServiceIdentityServiceCollectionExtensions.cs

```diff
@@ -8,24 +8,28 @@ namespace Microsoft.Extensions.DependencyInjection
     using Corvus.Identity.ManagedServiceIdentity.ClientAuthentication;
 
     /// <summary>
-    /// DI initialization for services using a Managed Service Identity (MSI).
+    /// DI initialization for services using a Managed Identity.
     /// </summary>
     public static class ServiceIdentityServiceCollectionExtensions
     {
         /// <summary>
-        /// Adds the Workflow Engine client to a service collection.
+        /// Adds an <c>AzureServiceTokenProvider</c>-based <see cref="IServiceIdentityTokenSource"/>
+        /// to a service collection.
         /// </summary>
         /// <param name="services">The service collection.</param>
+        /// <param name="options">Configuration for the token source.</param>
         /// <returns>The modified service collection.</returns>
-        public static IServiceCollection AddAzureMsiBasedTokenSource(
-            this IServiceCollection services)
+        public static IServiceCollection AddAzureManagedIdentityBasedTokenSource(
+            this IServiceCollection services,
+            AzureManagedIdentityTokenSourceOptions options = null)
         {
-            if (services.Any(s => s.ImplementationType == typeof(AzureMsiTokenSource)))
+            if (services.Any(s => s.ImplementationType == typeof(AzureManagedIdentityTokenSource)))
             {
                 return services;
             }
 
-            return services.AddSingleton<IServiceIdentityTokenSource, AzureMsiTokenSource>();
+            return services.AddSingleton<IServiceIdentityTokenSource>(
+                _ => new AzureManagedIdentityTokenSource(options?.AzureServicesAuthConnectionString));
```

Comment:
Seeing this in action, I think that this is a great way to deal with this problem.

One slight bump on the horizon that I see with this is that, as you aggregate these registrations up into the "top level" registrations (think the equivalent of `AddMVC()` - maybe `Menes.AddHttpRequestHosting()` or some such method) will we end up passing a lot of config through, or would we expect those top levels things to be given the config root and do that for themselves?

If the latter, how would you override individual instances? The same as usual by providing an override or fallback configuration provider when you build the config?

(I think I've answered my own questions.)

Reply:
One other idea @jesspanni and I talked about was whether to do something similar to a lot of the Microsoft libraries where you pass a callback during registration, and it calls you with an options type instance pre-populated with defaults, and which you can change.

For a leaf library like that, that doesn't really help with the issue you've raised, but it led to the realisation that the common pattern for Microsoft libraries seems to be that you have to pass a lot of config through with those too. If you look at things like the AAD auth config, they define options types but it's your problem to decide how to move stuff from config into those types. And it's OK. In fact, I like that it's very easy to see in one place what all your application's configuration settings are.

If, as we move up the stack, I see a couple of ways we could go. We could just introduce nesting into the options, e.g.:

```csharp
public class AstroArchitectureOptions
{
    public AzureManagedIdentityTokenSourceOptions ManagedIdentity { get; set; }
    public LeasingOptions Leasing { get; set; }
    public string SettingSpecificToThisLibrary { get; set; }
}
```

Applications can then consume this in one of two ways. They could just reflect the structure directly in their config, e.g.:

```json
{ "AstroArch": {
    "ManagedIdentity": { "AzureServicesAuthConnectionString": "Some=thing" },
    "Leasing": { "DefaultLeaseTime": "P1M" },
    "SettingSpecificToThisLibrary": "Spong"
  }
}
```

and then your DI init would look something like:

```csharp
services.AddAstroArch(configRoot.GetSection("AstroArch").Get<AstroArchitectureOptions>());
```

but if you wanted to split it up differently, you can change it fairly easily:

```csharp
var astroArchOptions = configRoot.GetSection("AstroArch").Get<AstroArchitectureOptions>();
astroArchOptions.Leasing = configRoot.GetSection("Leasing").Get<LeasingOptions>();
astroArchOptions.SettingSpecificToThisLibrary = runningInAzureFunction ? "Foo" : "Bar";
services.AddAstroArch();
```

This would mean that the `Leasing` section was not nested inside the `AstroArch` any more, but was being read from a separate top-level section from the config.

So this provides a fairly straightforward and transparent way to support "Just the usual", or "Basically the usual but I want this little bit done differently" scenarios without having to get into anything like crufting up an in-memory `IConfiguration` implementation.

And an alternative is, as you said, to solve this with that in-memory solution like we do in some parts today. We know that works. But I think the reason we ended up going down that path was only that we had to because lots of bits of code expect to be able to get hold of an `IConfiguration` or an `IConfigurationRoot` and read their settings out of there directly. If that's going on, then the custom layered config approach is the only way to get the flexibility. But my current view is that it would be better to get rid of all that deep-in-the-guts direct inspection of config, possibly even to the point of saying that out libraries shouldn't normally need a dependency on `Microsoft.Extensions.Configuration` at all.

Reply:
I agree with this nested options approach.

So, I think we need work items across the Corvus projects to enact this change systematically. It might be a good thing for everyone involved to pick up a WI and implement it as an "easy starter" item, but one that introduces a fundamental concept.

A side effect of embarking on that is that in the review process for those changes, it will afford a good opportunity to have a look at the DI registration methods across the board and see if they need any rationalisation; they rapidly get a bit spidery and horrible, and it is rarely clear what the expected "usual way in" is in the fact of a million registration methods that aggregate up.

[Just look at Aspnet MVC for a prime example of that at its apotheosis!]

 If we fan out a new set of work items for that, while doing review, we'll have hit two birds with one stone, and they can be picked up on the next pass through.